src/Processors/Sources/MongoDBSource.cpp

```diff
@@ -297,73 +353,77 @@ namespace
                     throw Exception(ErrorCodes::TYPE_MISMATCH, "Type mismatch, expected Array, got type id = {} for column {}",
                                     toString(value.type()), name);
 
-                size_t expected_dimensions = array_info[idx].num_dimensions;
-                const auto parse_value = array_info[idx].parser;
-                std::vector<Row> dimensions(expected_dimensions + 1);
-
-                auto array = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(value).value();
+                const auto nested_type = assert_cast<const DataTypeArray &>(data_type).getNestedType();
+                auto parent = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(value).value();
+                const size_t size = parent->size();
 
-                std::vector<std::pair<const Poco::MongoDB::Element *, size_t>> arrays;
-                arrays.emplace_back(&value, 0);
+                Field field = Array();
+                Array & arr = field.get<Array &>();
+                arr.reserve(size);
 
-                while (!arrays.empty())
+                for (size_t child_idx = 0; child_idx < size; ++child_idx)
                 {
-                    size_t dimension_idx = arrays.size() - 1;
+                    Poco::MongoDB::Element::Ptr child = parent->get(static_cast<int>(child_idx));
+                    arr.push_back(getValueField(nested_type, *child, name));
+                }
 
-                    if (dimension_idx + 1 > expected_dimensions)
-                        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Got more dimensions than expected");
+                assert_cast<ColumnArray &>(column).insert(field);
+                break;
```

Comment:
Working with Fields during insert is really suboptimal. Why don't just use `insertValue` function recursively here on nested column and Mongo array elements?